ChannelReaderActor 채널 변경 내역 전파 기능 추가

apptie opened the pull request:
## 📎 관련 Issue 번호
<!-- (필수) 해당 PR과 관련 있는 issue 번호를 입력해주세요. -->
- closed #29 
## 📄 작업 내용 요약
<!-- (필수) 작업한 내용을 간단히 요약해주세요. -->

### ChannelReaderActor 채널 변경 내역 전파 메시지 추가 

- ChannelReaderActor 외부에서 Notify 접두사 메시지를 받아 ClientSessionActor에게 Propagate 접두사 메시지 전달 

### ChannelReaderActor 전용 채팅 메시지 관리 컴포넌트 추가

- ChannelReaderActor 내에서 사용할 ChannelReaderChatMessages 컴포넌트 추가
- ChannelEntity 내에서 사용할 ChatMessages 네이밍을 명확하게 구분하기 위해 ChannelEntityChatMessages으로 네이밍 변경

apptie committed:
feat: ChannelReaderActor 전용 채팅 메시지 관리 컴포넌트 추가

apptie committed:
feat: ChannelReaderActor 채널 변경 내역 전파 메시지 추가

apptie committed:
refactor: ChannelReaderActor 전용 채팅 메시지 관리 컴포넌트 네이밍 변경

apptie committed:
refactor: ChannelEntity 전용 채팅 메시지 관리 컴포넌트 네이밍 변경

seonWKim commented on pekko/src/main/java/com/tok/pekko/domain/chat/actor/ChannelReaderChatMessages.java:
```diff
@@ -0,0 +1,273 @@
+    public ChatMessage update(Long messageId, String updatedMessage, LocalDateTime updatedAt) {
+        ChatMessageNode node = messageIdMap.get(messageId);
+
+        if (node == null) {
```
DB에는 존재하지만 evict된 메시지라서 `node == null`인 경우는 어떻게 처리되나요?

apptie replied:
그 상황은 전혀 고려하지 못했습니다
 
ChannelReaderChatMessages는 단순히 ChannelEntity로부터 이미 완료된 메시지를 전파받고 
최신 채팅 히스토리를 제한된 숫자만 관리하기 때문에 
이러한 상황에서 예외를 발생시키는것보다 무시하도록 하는 것이 자연스러울 것 같습니다 

ChannelReaderChatMessages와 ChannelEntityChatMessages 또한 이를 고려해서 수정하도록 하겠습니다

seonWKim commented on pekko/src/main/java/com/tok/pekko/domain/chat/actor/ChannelReaderChatMessages.java:
```diff
@@ -0,0 +1,273 @@
+        this.size = 0;
+    }
+
+    @ActorThreadSafe
```
모든 메서드들이 actor 내부에서 호출되는 거라면 class 레벨에 @ActorThreadSafe를 추가하는건 어떻게 생각하시나요?

apptie replied:
좀 더 간결하게 표현할 수 있어서 좋을 것 같습니다 
제가 미처 파악하지 못했다거나 잘못 판단해서 관련 없는 부분에 `@ActorThreadSafe`를 추가하는 걸 회피할 수도 있을 것 같습니다 

`@ActorThreadSafe`는 클래스, 메서드 레벨에 명시할 수 있도록 하고 
클래스 레벨에 붙이도록 변경하겠습니다

seonWKim replied:
나중에는 @ActorThreadSafe가 붙은 메서드나 클래스는 액터 클래스에서만 호출이되는지 컴파일타임에 검증하는 테스트 로직이 있어도 좋을 것 같아욤

apptie replied:
나중에 한 번 추가해보도록 하겠습니다

apptie committed:
refactor: @ActorThreadSafe 애노테이션을 클래스 레벨로 적용 가능하도록 변경

- 기존 메서드 레벨 적용 사례들을 클래스 레벨로 이전

apptie committed:
refactor: 메시지 삭제 시 ID만을 전파하도록 변경

- ChannelReaderActor에서 ClientSessionActor로 삭제된 메시지 ID만을 전달
- 웹소켓으로 삭제된 메시지 ID만을 전달

apptie committed:
refactor: 채팅 메시지 업데이트 시 ChatMessage 전체를 전파하지 않고 필요한 항목만 전파하도록 변경

- ChannelReaderActor에서 ClientSessionActor로 메시지 ID, 수정된 내용, 수정 시각만 전달
- 웹소켓으로 변경된 필드만 포함하도록 경량화

apptie committed:
refactor: ChannelReaderChatMessages 내부적으로 예외 상황을 무시하도록 변경

f-lab-devvy reviewed (approved):
LGTM 👍
